### ISSUE TITLE
Meta-issue: Different type assertion behaviors

### ISSUE THREAD

**Issue by RyanCavanaugh** [microsoft/TypeScript#26064]
This is a disambiguator issue for tracking various type assertion-like constructs that have been requested

### "is" type assertion

Tracking at #7481

**Description**: A type assertion operator that doesn't perform downcasting. This comes up reasonably often and is especially useful when you want to write an expression where the expression's *target* type is `any`/`unknown` but you know what type the expression *ought* to have.
```ts
// We want to validate that the object literal here is a Point3D,
// but the downcast here doesn't catch our miss of "z"
writeFile("point.json", JSON.stringify({ x: 2, y: 5 } _as_ Point3D);
```

This is also useful when writing function expressions or objects containing function expressions - its parameters can get contextually typed by the asserted-to type, but you really don't want downcasting of the return types to occur here.

#16605 tracks a different proposal for allowing a downcasting that downcasts the existence of missing properties but does not allow downcasting of types -- this is sort of ill-defined IMHO and needs some justification to exist as a separate entity.

### "pass-through" type assertion

Tracking at #26045

**Description**: You currently have to choose between index signature validation at the initialization site *or* getting type checking on property existence:
```ts
type Neat = { [key: string]: boolean };

const x: Neat = {
    m: true,
    n: 10 // Want error here...
};
// ... and here
console.log(x.z);
```

The idea is some kind of annotation that still allows for inference, but doesn't change the type of the expression

### Others

There are more to be found

**Comment by RyanCavanaugh** [comment]
TODO describe #10421

**Comment by agentcooper** [comment]
I'm finding the following behavior confusing:

```typescript
type Fruit = "apple" | "banana"
([1, 2, 3] as Array<[Fruit]>); // OK: error is expected
(["apple", "potato"] as Array<Fruit>); // Not OK: no error
```
[Plaground](https://www.typescriptlang.org/play/index.html#src=type%20Fruit%20%3D%20%22apple%22%20%7C%20%22banana%22%0D%0A(%5B1%2C%202%2C%203%5D%20as%20Array%3C%5BFruit%5D%3E)%3B%20%2F%2F%20OK%3A%20error%20is%20expected%0D%0A(%5B%22apple%22%2C%20%22potato%22%5D%20as%20Array%3CFruit%3E)%3B%20%2F%2F%20Not%20OK%3A%20no%20error).

Should this considered as a bug?
